Support intra-broker throttling (replica.alter.log.dirs.io.max.bytes.per.second)

dongjinleekr opened the pull request:
This PR resolves #1851.

dongjinleekr committed:
Add (default.)log.dir.throttle parameter

dongjinleekr committed:
Remove _throttleRate variable from ReplicationThrottleHelper class

dongjinleekr committed:
Rename ReplicationThrottleHelper method names: ThrottledRate → ReplicationThrottledRate

dongjinleekr committed:
Rename ReplicationThrottleHelper method name: setThrottles → setReplicationThrottles

dongjinleekr committed:
Add ReplicationThrottleHelper#LOG_DIR_THROTTLED_RATE support

- Add: setLogDirThrottles, setLogDirThrottledRateIfNecessary
- Rename: removeReplicationThrottledRateFromBroker → removeThrottledRatesFromBroker

dongjinleekr commented:
This PR is an WIP; I will test this feature in our in-house fork (@naver) this week. Stay tuned!! :upside_down_face:

dongjinleekr committed:
fix: #1851 Separate (un)setting throttling for inter/intra-broker rebalanacing

    - Rename: ReplicationThrottleHelper.clearThrottles → clearInterBrokerThrottles
    - Add ReplicationThrottleHelper.clearIntraBrokerThrottles
    - Executor.intraBrokerMoveReplicas now calls ReplicationThrottleHelper.setLogDirThrottles, ReplicationThrottleHelper.clearIntraBrokerThrottles

dongjinleekr commented:
@aswinayyolath @mhratson @jiao-zhangS Could you kindly have a look? We adopted this patch to our in-house distribution and confirmed it works correctly. :pray:

cpaika commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java:
```diff
@@ -1707,6 +1729,11 @@ private void intraBrokerMoveReplicas() {
         inExecutionTasks = inExecutionTasks();
       }
+
+      if (_logDirThrottle != null) {
```
for my understanding, why the underscore in the variable names here?

dongjinleekr replied:
In consistency with `_replicationThrottle`.

cpaika reviewed (approved):
Will make things a lot more stable for JBOD clusters, good change

dongjinleekr commented:
@cpaika @adkafka

Excuse me. Is there any issue with this PR? It initially worked on 2.5.138, and 2.5.141 was already released last month, but this PR has yet to be merged or released.

If any modifications are needed, don't hesitate to leave me a mention.

adkafka commented:
> @cpaika @adkafka
> 
> Excuse me. Is there any issue with this PR? It initially worked on 2.5.138, and 2.5.141 was already released last month, but this PR has yet to be merged or released.
> 
> If any modifications are needed, don't hesitate to leave me a mention.

I don't see any issues with this PR, but I'm not a maintainer of this project. We need a maintainer to review this and add their approval.

aswinayyolath commented:
@CCisGG could you please review this PR?

dongjinleekr commented:
@CCisGG Hello. Could you please have a look when you are free? :pray:

CCisGG commented:
Hi @mhratson, would you mind taking a look at this one? Thanks!

imans777 commented:
Anyone caring to review this?
It's really a must-have feature.

CCisGG commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/constants/ExecutorConfig.java:
```diff
@@ -206,6 +206,14 @@ public final class ExecutorConfig {
+   * <code>default.log.dir.throttle</code>
+   */
+  public static final String DEFAULT_LOG_DIR_THROTTLE_CONFIG = "default.log.dir.throttle";
+  public static final Long DEFAULT_DEFAULT_LOG_DIR_THROTTLE = null;
```
Would that be more reasonable to have it Long.MAX_VALUE rather than null? I'm a bit concerned to have a null default value for a LONG type. I'm not entirely sure what will happen here by default

CCisGG commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java:
```diff
@@ -830,7 +834,7 @@ public synchronized void executeProposals(Collection<ExecutionProposal> proposal
                             requestedBrokerLeadershipMovementConcurrency, requestedExecutionProgressCheckIntervalMs, replicaMovementStrategy,
                             isTriggeredByUserRequest, loadMonitor);
-      startExecution(loadMonitor, null, removedBrokers, replicationThrottle, isTriggeredByUserRequest);
+      startExecution(loadMonitor, null, removedBrokers, replicationThrottle, logDirThrottle, isTriggeredByUserRequest);
```
Just curious, does that make more sense to have a separate method for intra-broker throttling since this parameter only applies for inter-broker movement?

dongjinleekr replied:
I agree with you. But, since the `startExecution` method is invoked for all rebalance tasks yet (i.e., both inter-broker and intra-broker with demote, remove broker options), I thought the splitting task is better to be done with independent PR.

It is why I chose to add a new parameter here.

CCisGG replied:
Let's create a task to track the split. Thanks!

CCisGG commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java:
```diff
@@ -918,7 +922,7 @@ public synchronized void executeDemoteProposals(Collection<ExecutionProposal> pr
                             requestedClusterLeadershipMovementConcurrency, requestedBrokerLeadershipMovementConcurrency,
                             requestedExecutionProgressCheckIntervalMs, replicaMovementStrategy, isTriggeredByUserRequest, loadMonitor);
-      startExecution(loadMonitor, demotedBrokers, null, replicationThrottle, isTriggeredByUserRequest);
+      startExecution(loadMonitor, demotedBrokers, null, replicationThrottle, replicationThrottle, isTriggeredByUserRequest);
```
What is the implication here to have log dir throttling set to replication throttle here?

dongjinleekr replied:
Mistake; Since it is demote task, we don't need `logDirThrottle` parameter so it should be `null`. fixed.

CCisGG commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ReplicationThrottleHelper.java:
```diff
@@ -51,48 +52,54 @@ class ReplicationThrottleHelper {
-      for (Map.Entry<String, Set<String>> entry : throttledReplicas.entrySet()) {
-        setThrottledReplicas(entry.getKey(), entry.getValue());
-      }
+    LOG.info("Setting a rebalance throttle of {} bytes/sec", throttleRate);
```
Why do we remove the "if (throttlingEnabled())" check here?

dongjinleekr replied:
Before this PR, `clearThrottles` (now `clearInterBrokerThrottles`) had `if (throttlingEnabled())` check, which dependes on `private final Long _throttleRate;`, right?

```
private boolean throttlingEnabled() {
  return _throttleRate != null;
}
```

Since `ReplicationThrottleHelper` does not have `_throttleRate` anymore, the nullity check is moved into `Executor#interBrokerMoveReplicas` method:

```
# before
throttleHelper.setThrottles(tasksToExecute.stream().map(ExecutionTask::proposal).collect(Collectors.toList()));

# after
if (_replicationThrottle != null) {
  throttleHelper.setReplicationThrottles(tasksToExecute.stream().map(ExecutionTask::proposal).collect(Collectors.toList()),
      _replicationThrottle);
}
```

```
# before
throttleHelper.clearThrottles(completedTasks, inProgressTasks);

# after
if (_replicationThrottle != null) {
  throttleHelper.clearInterBrokerThrottles(completedTasks, inProgressTasks);
}
```

In short, the `if (throttlingEnabled())` check is moved from the body of `clearThrottles` to the outer if statement of the caller side, `Executor`.

CCisGG replied:
Got it. Thanks

CCisGG commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ReplicationThrottleHelper.java:
```diff
@@ -336,7 +355,7 @@ private void removeThrottledReplicasFromTopic(String topic, Set<String> replicas
   }
 
-  private void removeThrottledRateFromBroker(Integer brokerId)
+  private void removeThrottledRatesFromBroker(Integer brokerId)
```
Is this method now become "remove both throttle rates"? In that case we should add a comment to specify that.

dongjinleekr replied:
Oh no, it is just a simple mistake. There are no changes in the method body. fixed.

CCisGG reviewed (commented):
Added some comments/questions.

Some other questions I have:
1. Did you test that with your change, when no replication/intra broker throttling was set (which I think is by default), does this work properly?
2. Is their any case that this intra-broker throttling can interfere with inter-broker thorttling? E.g. what happens if they are enabled together?
3. Please update Configuration.md for this new config.

dongjinleekr committed:
Add 'default.log.dir.throttle' to Configurations.md

dongjinleekr committed:
Remove logDirThrottle parameter in Executor#executeDemoteProposals

dongjinleekr committed:
Revert method name: removeThrottledRatesFromBroker → removeThrottledRateFromBroker

dongjinleekr commented:
@CCisGG 

Sorry for being late. Here are the updates, rebased onto the latest main, and the anwsers:

> Did you test that with your change, when no replication/intra broker throttling was set (which I think is by default), does this work properly?

YES. I tested it before I opened the PR, and we have been running it with our in-house distribution for more than 6 months and experienced no trouble. (As you already pointed out, `default.log.dir.throttle` is `null` by default.)

> Is their any case that this intra-broker throttling can interfere with inter-broker thorttling? E.g. what happens if they are enabled together?

As far as I know, there is no case where intra and inter-broker throttlings interfere with each other. The reason is simple: the intra-broker rebalancing is only executed when `rebalance_disk` [parameter](https://github.com/linkedin/cruise-control/blob/main/cruise-control/src/main/resources/yaml/endpoints/rebalance.yaml#L144) is set to `true`. If this parameter is set to `false` (default), only inter-broker rebalancing is executed.

In short, intra- and inter-broker rebalancings are executed exclusively for each other, so if `default.replication.throttle` and `default.log.dir.throttle` are enabled together, they will not impact each other.

luciano-sabenca commented:
Any news about this MR? 
It's a really useful feature for us! 
:D

CCisGG commented on docs/wiki/User Guide/Configurations.md:
```diff
@@ -155,7 +155,8 @@ The following configurations are inherited from the open source Kafka client con
 | topics.excluded.from.partition.movement                            | String  | N         | ""                                                                                                                                                                                                                                                                                                                 ...
-| default.replication.throttle                                       | Long	  | N         | null	                                                                                                                                                                                                                                                                                                               ...
+| default.replication.throttle                                       | Long	| N         | null	                                                                                                                                                                                                                                                                                                                 ...
+| default.log.dir.throttle                                           | Long	| N         | null	                                                                                                                                                                                                                                                                                                                 ...
```
Please add different descriptions for both config since one of them is inter broker and the other is intra broker.

dongjinleekr replied:
OMG. What a silly mistake 😓 fixed.

CCisGG commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/KafkaCruiseControl.java:
```diff
@@ -651,6 +651,8 @@ private static boolean hasProposalsToExecute(Collection<ExecutionProposal> propo
    *                                (if null, use default.replica.movement.strategies).
    * @param replicationThrottle The replication throttle (bytes/second) to apply to both leaders and followers
```
Let's clarify in the comment that replicationThrottle is for inter-broker, logDirThrottle is for intra-broker rebalance.

dongjinleekr replied:
Fixed. I elaborated the Javadoc by explicitly stating inter-broker movements vs. intra-broker movements.

CCisGG reviewed (changes requested):
Thanks for the change! I left some minor comments.

dongjinleekr committed:
Fix documentation on the 'default.log.dir.throttle' configuration.

dongjinleekr committed:
Improve Javadoc of KafkaCruiseControl#executeProposals (i.e., inter-broker movements vs. intra-broker movements)

dongjinleekr committed:
Remove logDirThrottle parameter from KafkaCruiseControl#executeRemoval

dongjinleekr committed:
Remove _logDirThrottle from AddBrokersRunnable, RemoveBrokersRunnable, and UpdateTopicConfigurationRunnable. (reason: they do not incur any intra-broker replica movements)

dongjinleekr commented:
@CCisGG Rebased onto the latest main, with some additional improvements.

-----

> [Let's create a task to track the split. Thanks!](https://github.com/linkedin/cruise-control/pull/2145#discussion_r2002098925)

You mean splitting the `Executor#startExecution` method call to inter-broker mode and intra-broker mode, applying `replicationThrottle` and `logDirThrottle` respectively, right? I tried to, but it seems impossible. Here is why:

**The `Executor#executeProposals` method, which calls `Executor#startExecution`, is agnostic to the inter-broker and intra-broker modes.**

(Note: The method call chain is `KafkaCruiseControl#executeProposals` → `Executor#executeProposals` → `Executor#startExecution`)

The `rebalance_disk` parameter (`ParameterUtils.REBALANCE_DISK_MODE_PARAM`) is only used to generate the proposal, and then the proposal is delivered to the `Executor#executeProposals` method to be executed. That is, the `Executor#executeProposals` method can not discriminate between the two modes.

Then, is splitting `KafkaCruiseControl#executeProposals` into inter- and intra-broker modes feasible? I also investigated the possibility, but it seems impossible. Here is why: it is called by the following 5 API calls/Runnables:

1. `add_broker` (`AddBrokersRunnable`)
2. `fix_offline_replicas` (`FixOfflineReplicasRunnable`)
3. `rebalance` (`RebalanceRunnable`)
4. `RemoveDisksRunnable`
5. `UpdateTopicConfigurationRunnable`

5 incurs no movements, 1 incurs inter-broker movements only, 3 incurs either inter-broker or intra-broker movements, and 2 and 4 incur BOTH. For this reason, splitting the two methods' calls seems unreasonable - we can not assert whether 2 or 4 invocations need intra-broker movements beforehand.

For the reasons above, I did nothing to the call chain. It is true that these methods seem too overloaded, but improving them with another PR may be much better IMHO. (I have an idea, and will open another PR on this issue after this PR is merged.)

dongjinleekr commented:
@CCisGG Instead, I did:

1. removed `logDirThrottle` from the `Executor#executeRemoval` method.
    Since its goal is to move all replicas in the specified broker, intra-broker replica movement does not occur at all. So, it is totally useless.
2. Remove `_logDirThrottle` from the AddBrokersRunnable, RemoveBrokersRunnable, and UpdateTopicConfigurationRunnable.
    DITTO. These Runnables do not incur any intra-broker replica movements. So, it is also useless.

With one commit for `Configurations.md`, one commit for `KafkaCruiseControl` Javadoc, and two commits for the improvements above. In total, I added four commits after your last review. :pray: Please have a look when you are free!! Thanks.

CCisGG commented:
@dongjinleekr Thanks for the explanations and the new commits. I think they are all reasonable. 

Since I probably won't have much time to do another round of review and audit, would you mind inviting some other person (maybe from your org) to review the PR? And Let's do a final round of testing to ensure this won't interfere with inter-broker rebalance. Once those are done, I can help to merge to PR. Let me know if this works. Thank you!

CCisGG reviewed (approved):
Looks good in general to me. Please ask for another review from a different person if possible. Then I can help to merge the PR.

dongjinleekr commented:
@CCisGG 

> Would you mind inviting some other person (maybe from your org) to review the PR?

Oh no, no problem. In fact, I am now working on supporting this feature in the [Strimzi Kafka operator](https://github.com/strimzi/strimzi-kafka-operator), and it will be completed by this weekend. As soon as I open the PR, there may be other people who can review it. (Sadly, I am the only person who is working on CC internals in my org :cry:)

Plus: We are running this feature in our in-house fork but, let me double-check if there is any problem 🙏 Thank you!

kyguy commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/handler/async/runnable/UpdateTopicConfigurationRunnable.java:
```diff
@@ -209,8 +209,8 @@ protected OptimizerResult workWithClusterModel() throws KafkaCruiseControlExcept
           0, _clusterLeaderMovementConcurrency, _brokerLeaderMovementConcurrency,
-          _executionProgressCheckIntervalMs, _replicaMovementStrategy, _replicationThrottle, _isTriggeredByUserRequest, _uuid,
-          SKIP_AUTO_REFRESHING_CONCURRENCY);
+          _executionProgressCheckIntervalMs, _replicaMovementStrategy, _replicationThrottle, null,
```
The `logDirThrottle` is null here because partitions are never any intra-broker partition movements as part of topic configuration changes, is my understanding correct?

kyguy commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/handler/async/runnable/AddBrokersRunnable.java:
```diff
@@ -127,6 +127,7 @@ protected OptimizerResult workWithClusterModel() throws KafkaCruiseControlExcept
                                            _executionProgressCheckIntervalMs,
                                            _replicaMovementStrategy,
                                            _replicationThrottle,
+                                           null,
```
Why is the `logDirThrottle` argument set to null here? Wouldn't we want new brokers to have the configured logDirThrottle set?

kyguy reviewed (commented):
Hi @dongjinleekr, the PR looks great overall! I just left a few minor suggestions, mostly related to refactoring for consistency and readability.

palpandy commented:
Hello, 
Any news about this feature ? It is really a important feature for us.
